GraphStreamingProject/GraphZeppelin
# Improvements to Connected Components Performance

## Big Changes
- We now can either backup supernodes to disk or in memory based upon a configuration flag set by the user
- Made Boruvka emulation much more parallel by serially computing the DSU changes and tracking the merges that will need to happen. Then in parallel performing the merges. The boruvka emulation now consists of the following three stages performed repeatedly.
  1. Query Active Supernodes (parallel)
  2. Perform DSU updates and identify merging to be done (serial)
  3. Merge the Supernodes (parallel)
- Additionally, changed the backing up of supernodes to only occur when the supernode in question has been selected by the DSU stage for merging during the first round of Boruvka

## Minor Changes
#### Configuration
- Renamed 'path_prefix' in configuration options to 'disk_dir' to make it clear what it does.
- 'disk_dir' is now where both the gutter tree and Supernodes backed up to disk live

#### Connected Components Algorithm
- Made Graph::update inline (maybe minor performance benefit)
- Added ability to reset Supernode idx and Sketch already_queried variables (useful for continuous CC)
- Renamed function that runs the CC algorithm to 'boruvka_emulation'
- Made all CCs route through Graph::connected_components(bool) and made that the only function that calls BufferingSystem::force_flush()

#### Sketching
- Fixed bug in sketching (we need to use 64 bit column hashes as (2^32)^2 = 2^64)
- Changed contains query to use bitwise & rather than modulus as modulus would break on 2^32 nodes (as we would need to check if all 64 bits are 1 and therefore would attempt to mod by 2^65.

#### Testing
- Fixes graph tests to not allow for failures as these failures should not happen whp
- Add a test for queries during the stream that backs up in memory and on disk
- Add a test that uses parallel processing
